### PR TITLE
add rabbit hole post

### PR DESCRIPTION
**Pull request by mediocregopher**
An engaging post about a rabbit looking to find his place in a big rabbit world. Join him as he finds danger, adventure, and maybe even a little love. A fun-loving post that the whole family will enjoy.

(Actually it's about me debugging some shit)

### CODE REVIEW

File: _posts/2015-3-11-rabbit-hole.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+extensions in place, all of them written in C and all of them essentially black
+boxes; any of them could be the culprit.  Grepping through the likely
+candidates' source code for the string `unknown` proves fruitless.
+
+We try xdebug at this point. xdebug is a profiler for php which will create
+cachegrind files for the code that's being run. With cachegrind you can see
+every function which was ever called, how long spent within each function, a
+full call-graph, and lots more. Unfortunately xdebug outputs cachegrind files on
+a per-php-fpm-process basis, and overwrites the previous file on each new
+request. So xdebug is pretty much useless, since we can't know that what is
```
Comment (MarcoPolo) [inline _posts/2015-3-11-rabbit-hole.md]: phrasing is a bit awkward: `can't know that what is`

**Comment by MarcoPolo** [comment]
LGTM

File: _posts/2015-3-11-rabbit-hole.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+Our setup on the FENs involves users hitting nginx for static content and nginx
+proxying PHP requests back to php-fpm. Since all of our virtual domains are
+defined individually in nginx, we are able to do something horrible.
+
+On the particular FEN we're on we make a guess as to which virtual domain the
+problem is likely coming from (our main app), and proxy all traffic from all
+other domains to a different FEN. We still see requests for `unknown` leaving
+the box, so we've narrowed the problem down a little more.
+
+## The Dispair
```
Comment (theabraham) [inline _posts/2015-3-11-rabbit-hole.md]: despair

Comment (mediocregopher) [inline _posts/2015-3-11-rabbit-hole.md]: TIL

Comment (theabraham) [inline _posts/2015-3-11-rabbit-hole.md]: maybe that's the Canadian/British spelling

**Comment by mediocregopher** [comment]
Fixed the things marco and ibrahim pointed out

**Comment by fastest963** [comment]
LGTM

File: _posts/2015-3-11-rabbit-hole.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+`$_SERVER['REMOTE_ADDR']`, which is the variable populated with the IP address
+of the client, is sometimes `unknown`. We guess that this has something to do
+with some magic we are doing on nginx's side to populate `REMOTE_ADDR` with the
+real IP address of the client in the case of them going through a proxy.
+
+Many proxies send along the header `X-Forwarded-For` to indicate the real IP of
+the client they're proxying for, otherwise the server would only see the proxy's
+IP. In our setup I decided that in those cases we should set the `REMOTE_ADDR`
+to the real client IP so our application logic doesn't even have to worry about
+it. There a couple problems with this which render it a bad decision, one being
```
Comment (likethemammal) [inline _posts/2015-3-11-rabbit-hole.md]: "There a couple" should be "There are a couple".

File: _posts/2015-3-11-rabbit-hole.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+that if some misbahaving proxy was to, say, start sending `X-Forwarded-For:
+unknown` then some written applications might mistake that to mean the client's
+IP is `unknown`.
+
+## The Fix
+
+The fix here was two-fold:
+
+1) We now always set `$_SERVER['REMOTE_ADDR']` to be the remote address of the
+requests, regardless of it it's a proxy, and also send the application the
```
Comment (likethemammal) [inline _posts/2015-3-11-rabbit-hole.md]: "It it's " should be "if it's"

File: _posts/2015-3-11-rabbit-hole.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
+---
+layout: post
+title: Rabbit Hole
+tags: debug php sysdig dns
+username: mediocregopher
+---
+
+# Rabbit Hole
+
+We've begun rolling out [SkyDNS][skydns] at my job, which has been pretty neat.
```
Comment (akenn) [inline _posts/2015-3-11-rabbit-hole.md]: I had no idea you could link to github repos like this!

File: _posts/2015-3-11-rabbit-hole.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 65535 bytes
+    ...
+    $fen_ip.50257 > $skydns_ip.domain: [udp sum ok] 16218+ A? unknown. (25)
+    $fen_ip.27372 > $skydns_ip.domain: [udp sum ok] 16218+ A? unknown. (25)
+    $fen_ip.35634 > $skydns_ip.domain: [udp sum ok] 59227+ A? unknown. (25)
+    $fen_ip.64363 > $skydns_ip.domain: [udp sum ok] 59227+ A? unknown. (25)
+```
+
+It appears that some of our front end nodes (FENs) are making tons of DNS
+requests trying to find the A record of `unknown`. This is not going to work;
```
Comment (akenn) [inline _posts/2015-3-11-rabbit-hole.md]: I think you can remove "This is not going to work;"

File: _posts/2015-3-11-rabbit-hole.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+something on our FENs is doing something insane and is breaking.
+
+## The FENs
+
+Hopping over to my favorite FEN we're able to see the packets in question
```
Comment (akenn) [inline _posts/2015-3-11-rabbit-hole.md]: By favorite do you just mean a random FEN? If so I'd just say "Hopping on to one of our FENs we're able to see".

File: _posts/2015-3-11-rabbit-hole.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+leaving on a tcpdump as well, but that's not helpful for finding the root cause.
+We have lots of processes running on the FENs and any number of them could be
+doing something crazy.
+
+We fire up sysdig, which is kind of like systemtap (which is kind of like
+strace), but easier to use. The following command gets us what we want to know:
```
Comment (akenn) [inline _posts/2015-3-11-rabbit-hole.md]: To dig a little deeper we'll fire up sysdig, a tool for capturing and analyzing system activity (it' kind of like systemtap and strace, but easier to use). The following command will show us in realtime all sockets with udp as their IP protocol:

Comment (akenn) [inline _posts/2015-3-11-rabbit-hole.md]: If you want you can keep the first sentence as-is. I just think a short explanation of what the command is doing would be nice.

File: _posts/2015-3-11-rabbit-hole.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+The fix here was two-fold:
+
+1) We now always set `$_SERVER['REMOTE_ADDR']` to be the remote address of the
+requests, regardless of it it's a proxy, and also send the application the
+`X-Forwarded-For` header to do with as it pleases.
+
+2) Inside our app we look at all the headers sent and do some processing to
+decide what the actual client IP is. PHP can handle a lot more complex logic
+than nginx can, so we can do things like check to make sure the IP is an IP, and
+also that it's not some NAT'd internal ip, and so forth
```
Comment (akenn) [inline _posts/2015-3-11-rabbit-hole.md]: Period at the end.

File: _posts/2015-3-11-rabbit-hole.md (hunk 5 above)
Comment (akenn) [inline _posts/2015-3-11-rabbit-hole.md]: Oh nvm I see the link at the bottom *derp*

File: _posts/2015-3-11-rabbit-hole.md (hunk 7 above)
Comment (mediocregopher) [inline _posts/2015-3-11-rabbit-hole.md]: Nope, we play favorites :P Mine was RHL060 till its lease went up and it went back, nowadays I'm partial to RHL206

**Pull request merged**